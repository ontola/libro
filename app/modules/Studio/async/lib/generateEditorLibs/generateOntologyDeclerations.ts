
import { isNamedNode } from '@ontologies/core';

import { validVariableRegex } from './validVariableNameRegex';

export type OntologyMap = Array<[Record<string, unknown>, string]>;

const buildOntologyDeclaration = (ont: Record<string, unknown>, name: string) => {
  const keys = Object.keys(ont);
  const mappedKeys = keys.filter((key) => isNamedNode(ont[key]))
    .map((key) => {
      const isRegularVariableName = validVariableRegex.test(key);
      const rightHand = isRegularVariableName ? key : `['${key}']`;

      return `${rightHand}: NamedNode;`;
    });

  return [
    `declare type ${name.toUpperCase()} = {`,
    'ns: (term: string) => NamedNode,',
    ...mappedKeys,
    '};',
    `declare const ${name}: ${name.toUpperCase()};`,
  ].flat().join('\n');
};

export const generateOntologyDeclerations = (ontologyMap: OntologyMap): string => [
  ...ontologyMap.map((ont) => buildOntologyDeclaration(ont[0], ont[1])),
].flat().join('\n');
