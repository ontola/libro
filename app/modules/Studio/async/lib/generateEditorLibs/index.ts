import * as schema from '@ontologies/schema';
import * as rdfx from '@ontologies/rdf';
import * as rdfs from '@ontologies/rdfs';
import * as as from '@ontologies/as';

import sales from '../../../../../ontology/sales';
import ontola from '../../../../../ontology/ontola';
import argu from '../../../../../ontology/argu';
import elements from '../../../../../ontology/elements';
import dbo from '../../../../../ontology/dbo';
import fa4 from '../../../../../ontology/fa4';
import appSlashless from '../../../../../ontology/appSlashless';
import form from '../../../../../ontology/form';
import libro from '../../../../../ontology/libro';
import link from '../../../../../ontology/link';
import meeting from '../../../../../ontology/meeting';

import { OntologyMap, generateOntologyDeclerations } from './generateOntologyDeclerations';

const extraLibs = [
  'declare type NamedNode = string;',
  'declare function seq<T = any>(arr: T[], id?: string | undefined): Object;',
];

export const generateEditorLibs = (): string => {
  const onts: OntologyMap = [
    [ontola, 'ontola'],
    [sales, 'sales'],
    [argu, 'argu'],
    [elements, 'elements'],
    [dbo, 'dbo'],
    [rdfx, 'rdfx'],
    [rdfs, 'rdfs'],
    [schema, 'schema'],
    [fa4, 'fa4'],
    [appSlashless, 'appSlashless'],
    [form, 'form'],
    [libro, 'libro'],
    [link, 'link'],
    [meeting, 'meeting'],
    [as, 'as'],
  ];

  const declarations = [
    ...extraLibs,
  ].flat().join('\n');

  return `${declarations}\n${generateOntologyDeclerations(onts)}`;
};
