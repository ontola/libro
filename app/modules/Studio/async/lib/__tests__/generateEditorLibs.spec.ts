import { createNS } from '@ontologies/core';

import { OntologyMap, generateOntologyDeclerations } from '../generateEditorLibs/generateOntologyDeclerations';
import { generateEditorLibs } from '../generateEditorLibs/index';

const ns = createNS('https://fuckblockchain.com/ns#');

const fakeOntology = {
  ns,
  // eslint-disable-next-line sort-keys
  TrendGraph: ns('TrendGraph'),
  ['Weird:Variable//Thing']: ns('Weird:Variable//Thing'),
  trend: ns('trend'),
};

const ontologyMap: OntologyMap = [
  [fakeOntology, 'fakeOntology'],
];

describe('generateEditorLibs', () => {
  it('should return a string', () => {
    const result = generateEditorLibs();
    expect(typeof result).toBe('string');
  });

  describe('generateOntologyDeclerations', () => {
    it('should return a string', () => {
      const result = generateOntologyDeclerations([]);
      const result2 = generateOntologyDeclerations(ontologyMap);

      expect(typeof result).toEqual('string');
      expect(typeof result2).toEqual('string');
    });

    it('should parse the ontology correctly', () => {
      const result = generateOntologyDeclerations(ontologyMap);
      expect(result).toMatchSnapshot();
    });
  });
});
