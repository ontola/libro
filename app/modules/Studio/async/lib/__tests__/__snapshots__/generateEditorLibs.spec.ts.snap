// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateEditorLibs generateOntologyDeclerations should parse the ontology correctly 1`] = `
"declare type FAKEONTOLOGY = {
ns: (term: string) => NamedNode,
TrendGraph: NamedNode;
['Weird:Variable//Thing']: NamedNode;
trend: NamedNode;
};
declare const fakeOntology: FAKEONTOLOGY;"
`;
